### ISSUE TITLE
FYI: emacs' new native line numbers break company's completion popup

### ISSUE THREAD

**Issue by Alexander-Miller** [company-mode/company-mode#678]
In case you haven't seen it on emacs-devel yet:

 * [General discussion](https://lists.gnu.org/archive/html/emacs-devel/2017-06/msg00318.html)
 * [Company-specific bug report](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=27427)

You can also reproduce the problem without the new line numbers by using `(setq line-prefix "......")`

**Comment by dgutov** [comment]
Thanks for the report.

> You can also reproduce the problem without the new line numbers by using (setq line-prefix "......")

Could you show how *that* looks on your machine? In my testing, the result is also broken, but significantly different from the image you posted to debbugs (http://imgur.com/xHf2mlp).

**Comment by Alexander-Miller** [comment]
Added pictures to the bug report discussion.

You know all this talk of popup troubles gave me a weird idea. Why not hook up some external application in lieu of an emacs-native popup? Just feed it completion candidates in stdin and have it write back selection in stdout. Mostly this is about me finding a reason to learn and use clojure (if it can manage to start quickly enough), but this might give us something async, and easily rendered and maybe even nice looking.
I figured I'd ask your opinion first just in case I'm missing some catch somewhere that makes this idea much more difficult if not outright impossible.

Just as I was typing this I remembered that this is pretty much exactly what rofi does, and I know and use rofi myself. That should make getting a prototype much easier.

**Comment by dgutov** [comment]
> Why not hook up some external application in lieu of an emacs-native popup? Just feed it completion candidates in stdin and have it write back selection in stdout.

For one thing, there would have to be such application on all supported platforms.

Second, it's hard to do while keeping the existing features. For instance, now you can type, see the popup appear automatically, but ignore it and continue interacting with Emacs, as long as you don't use one of the limited set of key combinations.

You are probably thinking more of an alternative UI along the lines of `helm-company` which is always invoked explicitly. You could write a similar one with an external chooser program.

Not sure how it would help with anything "async", though.

**Comment by dgutov** [comment]
> Added pictures to the bug report discussion

Thanks!

**Comment by Alexander-Miller** [comment]
> You are probably thinking more of an alternative UI along the lines of helm-company

Exactly, I didn't mean to make the thing a part of official emacs. I just meant something that I can do on my own, for fun and profit, and *maybe*, if it turns out to actually work well enough, offer it as a package, something you can use as an alternative company frontend.

> Not sure how it would help with anything "async", though.

I guess emacs could be continously doing work to search for completion candidates, while the popup program remains responsive while accepting, and selecting and filtering what it's fed by emacs. Don't know if would really do any good, it's just something that came to mind since I've done some work with async stuff recently.

**Comment by dgutov** [comment]
> something you can use as an alternative company frontend

Cool. Just to be clear, though, it wouldn't fit the API of `company-frontends`.

> while the popup program remains responsive while accepting

That sounds iffy. The workflow I'm imagining has the user invoking completion command explicitly, because they are going to see an external program that only serves to show and filter completions. Might as well perform completion synchronously.

Maybe you are thinking of streaming, though. But `company-backends` can't do that, so far.

**Comment by Alexander-Miller** [comment]
> Cool. Just to be clear, though, it wouldn't fit the API of company-frontends.

Oh well, dirty workaround it is, then.
I've looked at the docs of `company-frontends` and it looks doable at first glance. Which part would not fit?

>  they are going to see an external program that only serves to show and filter completions. Might as well perform completion synchronously.

Fair enough. Though the external popup needn't be for showing only. If it's something I have complete control over, unlike rofi,  I could simply send all key input back to emacs, so that for example you could still "type in emacs" while the popup is open.

**Comment by dgutov** [comment]
> Which part would not fit?

Like, all of it. You can't pop up the external program and switch focus to it in the `post-command` handler. Again, see https://github.com/company-mode/company-mode/issues/678#issuecomment-310142387.

`helm-company` is not a company-frontend as well.

> If it's something I have complete control over, unlike rofi, I could simply send all key input back to emacs

I'd like to see that.

**Comment by Alexander-Miller** [comment]
> Like, all of it.

Ok, I've never interacted with this API, I'll see if I can find some alternative solution.

> I'd like to see that.

I already have some dead simple elisp that exchanges a Hello World with a python script a few times and  runs insert on the answer, so it's possible and easy enough. I assume you mean an actual working popup that also knows how to deal with complex keypress combinations. I'll make sure to drop by again if I really do get something working.

**Comment by Alexander-Miller** [comment]
[There's a working rofi prototype now](https://gist.github.com/Alexander-Miller/1ca7ccc9a53754084c998662824c86ab). Having used it a bit while doing actual elisp programming it works surprisingly ok. If you want to try it yourself:
 * You need an up-to-date rofi - a colleague on ubuntu got 0.15 from the official repos while the current release is 1.3.1
 * Rofi has focus issues on wayland, you won't be able to type or select anything

**Comment by dgutov** [comment]
> Ok, I've never interacted with this API, I'll see if I can find some alternative solution.'

You can check out how e.g. `helm-company` does this.

> I assume you mean an actual working popup that also knows how to deal with complex keypress combinations.

Yes, of course. That's the hard part.

> There's a working rofi prototype now

Cool. The screenshot looks nice, but I'd rather not spend time on fixing the dependencies to build it (1.2.0 from Ubuntu's repo didn't work).

But this kind of display presents a problem, I think, even if you can make it pass keyboard events back to Emacs: it hides more than a third of its window. For it to work more or less unobtrusively, it would have to take up less space as well.

**Comment by Alexander-Miller** [comment]
> You can check out how e.g. helm-company does this.

Will do.

> Yes, of course. That's the hard part.

And will have to wait until I actually have a popup of my own going. Saving best for last.

> it hides more than a third of its window. For it to work more or less unobtrusively, it would have to take up less space as well.

All we have now is a proof of concept - usability not included. I just used my default settings for rofi, that's what it looks like when I use rofi as a program launcher. The kind of looks I have in mind would be no different that what you'd expect to see in Eclipse or Atom. Unfortunately I cannot demonstrate proper placement with rofi, its spawn location is limited to up/down/left/right settings.

I've driven this issue quite a bit off topic now, so fell free to close it if you don't want to keep it for its original topic. I've got all the info I need now, and will let you know if and when I do get something working.
